rp2040-hal/examples/uart.rs

```diff
@@ -0,0 +1,70 @@
+//! Print an incrementing number to UART0 in a loop
+#![no_std]
+#![no_main]
+
+use core::fmt::Write;
+use cortex_m_rt::entry;
+use hal::clocks::init_clocks_and_plls;
+use hal::gpio::{self, Pins};
+use hal::pac;
+use hal::sio::Sio;
+use hal::uart::UartPeripheral;
+use hal::watchdog::Watchdog;
+use panic_halt as _;
+use rp2040_hal as hal;
+
+#[link_section = ".boot2"]
+#[used]
+pub static BOOT2: [u8; 256] = rp2040_boot2::BOOT_LOADER;
+
+#[entry]
+fn main() -> ! {
+    let mut pac = pac::Peripherals::take().unwrap();
+    let mut watchdog = Watchdog::new(pac.WATCHDOG);
+    let sio = Sio::new(pac.SIO);
+
+    // External high-speed crystal on the pico board is 12Mhz
+    let external_xtal_freq_hz = 12_000_000u32;
+    let clocks = init_clocks_and_plls(
+        external_xtal_freq_hz,
+        pac.XOSC,
+        pac.CLOCKS,
+        pac.PLL_SYS,
+        pac.PLL_USB,
+        &mut pac.RESETS,
+        &mut watchdog,
+    )
+    .ok()
+    .unwrap();
+
+    let pins = Pins::new(
+        pac.IO_BANK0,
+        pac.PADS_BANK0,
+        sio.gpio_bank0,
+        &mut pac.RESETS,
+    );
+
+    let mut uart = UartPeripheral::<_, _>::enable(
+        pac.UART0,
+        &mut pac.RESETS,
+        hal::uart::common_configs::_9600_8_N_1,
+        clocks.peripheral_clock.into(),
+    )
+    .unwrap();
+
+    // UART TX (characters sent from pico) on pin 1 (GPIO0) and RX (on pin 2 (GPIO1)
+    let _tx_pin = pins.gpio0.into_mode::<gpio::FunctionUart>();
+    let _rx_pin = pins.gpio1.into_mode::<gpio::FunctionUart>();
+
+    // We need a short delay here otherwise the first few characters are garbled.
+    // TODO: work out why
```

Comment:
This is probably due to missing frame start. While the UART channel is not in use, the line should be in marking state, i.e. high, but the above lines turn the line from high impedance state to high state and hence the line might not be in the marking state for long enough for the receiver to correctly detect this. Maybe you could initialize the pin as output before initializing the UART peripheral and see whether that helps?

Reply:
Ah, of course.
I was observing this while testing by physically detaching and re-attaching the Pico while leaving my USB-UART attached, so it would make sense that TX wasn't in a good state prior to transmission.
It isn't exhibiting this issue when resetting the program via a debugger or by via the RUN pin, so I think it's safe to remove this delay.